README.txt
s3DAV
=====
s3DAV is a webDAV server which allows you to access a virtual file system using your Amazon s3 account.

How to run
==========
Just type in this command :
java -jar s3DAV-0.2.jar

This will open a s3DAV server on the port 8070.

The configuration page is : http://127.0.0.1:8070/index.html

More information
================
More information can be found here: http://www.carion.org/s3dav/index.html

Support
=======
Questions should be sent to : pcarion@gmail.com

CHANGELOG.txt
## Changelog
### Release 0.7.1
* FIX: the S3 keys where not properly updated when entered from the admin console

### Release 0.7
* NEW: s3DAV is now also a ftp server (default port is 21). 

* NEW: files being uploaded appear in the list of files with a size set to 1. You can see
  the list of current uploads here: http://127.0.0.1:8060/index.html?page=uploads

* FIX: due to a bug on Memory Mapped file under windows (http://bugs.sun.com/bugdatabase/view_bug.do?bug_id=4715154)
  we don't use anymore Memory Mapped file
  
* NEW: command line arguments to set the s3 account information

* NEW: new location for authorization file (%user.home%/s3dav/authorization). No need to reauthenticate, the previous
  authorizationfile will be copied in this new location.

* FIX: fixes in the webDAV protocol implementation to pass the "litmus" tests
  (http://www.webdav.org/neon/litmus/). litmus - WebDAV server protocol compliance test suite
  The only tests which fail (3 on 41) are related to the PROPPATCH command: I've chosen not
  to implement this command, as it's not needed to copy files into a S3 account.
  
* NEW: you can set the value of the port to be used when launching s3Dav. For, instance, you can 
  start s3DAV with java -jar s3DAV-x.y.jar 8060 8070 21. The first port is the admin console, the seconde 
  one is the webDAV server, the last one is the ftp port.

### Release 0.6

* FIX: change the InputStream management.
  s3DAV has now been tested with DAV Explorer
  (http://www.ics.uci.edu/~webdav/download.html).
  Thank you Laurian G. for telling me about DAV Explorer.

* NEW: there is now a cache mechanism used to cache the 'HEAD' requests.
  This should improve the performances as a lot of 'HEAD' request are
  done to get the status of the file system.

* FIX: you can now create directories and files with spacs or & in their
  name

* NEW: in the "buckets page", from the "raw listing" link,
  you can now list *all* the objects sored in your bucket.
  You'll see there any object, even those not stored through s3DAV and
  which would then not being 'seen' by the s3DAV file system

* FIX: on webdav delete, returns 200 (OK) instead of 404 is the object
  does not exist. This will fix the problem if you move a directory
  isnide your s3 file system from Windows explorer.

### Release 0.5 (not uploaded under sourceforge)
* FIX: log files were not rotating properly.

### Release 0.4
* NEW: there is a new page in the admin page ( http://127.0.0.1:8070 ) called =Logs Page=.
  This page allows you to see the logs (which are also sent to the standard output)

* NEW: logs are stored in a rotating file located in : =${user.home}/s3dav/log/log{0,1}.txt=

* FIX: in the browse view (HTML), the name are now properly displayed.

* NEW: the logs have been changed in order to help tracking of problems

* FIX: a double encoding when getting the parent URI made impossible to
  create a directory containing a space, if one of the ancestor already had
  a space in its name.

* FIX: a HEAD request on a XML file has a content-type text/xml and this would trigger
  the reading of the body of the response. Now, we never read the body on HEAD requests.

### Release 0.3

* NEW: On the Website, there is now a FAQ page. Very limited now, the main
  purpose of this page was to discuss the naming convention used by s3DAV to
  define the keys. This explains also why, for now, you can't see the content
  uploaded by a different tool than s3DAV

* NEW: You can browse the content from your Internet Browser by
  accessing the URL http://127.0.0.1:8070

* NEW: when copying file into your s3 account, the content
  is no more fully stored into memory: it's copied into a file
  which is then mapped into memory.

* FIX: leading and trailing spaces are removed from
  the ids in the AWS account form.

* FIX: when we get content from the S3 repository (object GET)
  the content is fully read in memory before being sent back to the
  webDAV request.
  Now, the InputStream is directly sent back to the initiating webDAV
  request. The connection with S3 is closed when all the data have been
  read from the webDAV request.

* NEW: the version number of s3DAV is displayed on startup

* NEW: in the S3 requests, the USer-Agent is now set to =s3DAV/<version number>=

* FIX: the server is not working properly when compiled with JDK 1.5 - we get
  a 403 error back from the server.
  The reason is that, when sending a HEAD request, the =HttpURLConnection= class is
  automatically adding the content type =application/x-www-form-urlencoded= and this
  breaks the Authorization mechanism.

  Now, when a content-type is not set by the application, we always set the content-type to "application/x-www-form-urlencoded",
  even though there is no content associated to the message.

### Release : 0.2


src/org/carion/s3dav/Main.java
/*
 * Copyright (c) 2006, Pierre Carion.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.carion.s3dav;

import java.io.File;

import org.carion.s3.Credential;
import org.carion.s3.S3Log;
import org.carion.s3.admin.AdminServer;
import org.carion.s3.impl.CredentialFactory;
import org.carion.s3.impl.S3RepositoryImpl;
import org.carion.s3.util.LogWriter;
import org.carion.s3.util.S3LogImpl;
import org.carion.s3dav.webdav.WebdavServer;
import org.carion.s3ftp.FtpServer;

public class Main {

    public static void main(String[] args) {
        String userHome = System.getProperty("user.home");
        File s3DavDir = new File(userHome, "s3dav");
        if (!s3DavDir.isDirectory() && !s3DavDir.mkdirs()) {
            throw new RuntimeException("Can't create directory:" + s3DavDir);
        }

        // check if user wants to create credentials
        if ((args.length == 3) && ("-x".equals(args[0].trim()))) {
            createCredentials(s3DavDir, args[1], args[2]);
            System.exit(0);
        }

        int adminServerPort = getPortValue(args, 0, 8060);
        int webdavServerPort = getPortValue(args, 1, 8070);
        int ftpServerPort = getPortValue(args, 2, 21);

        try {
            // 1) Initialize log mechanism
            File logDir = new File(s3DavDir, "log");
            if (!logDir.isDirectory() && !logDir.mkdirs()) {
                throw new RuntimeException("Can't create directory:" + logDir);
            }
            LogWriter logWriter = new LogWriter(true, logDir);
            S3Log log = new S3LogImpl(logWriter);
            log.log("s3DAV - version:" + Version.VERSION);

            // 2) Initialize s3 repository access
            Credential credential = CredentialFactory.getCredential(s3DavDir);

            File uploadDir = new File(s3DavDir, "upload");
            if (!uploadDir.isDirectory() && !uploadDir.mkdirs()) {
                throw new RuntimeException("Can't create directory:"
                        + uploadDir);
            }

            S3RepositoryImpl repository = new S3RepositoryImpl(credential,
                    s3DavDir, uploadDir, log.getLogger(">s3>"));

            // 3) Initialize admin server
            if (adminServerPort > 0) {
                AdminServer adminServer = new AdminServer(adminServerPort,
                        repository, logWriter, repository.getUploadManager(),
                        log.getLogger(">admin>"));
                adminServer.start();
            }

            // 4) Initialize s3DAV server
            if (webdavServerPort > 0) {
                WebdavServer webdavServer = new WebdavServer(webdavServerPort,
                        repository, log.getLogger(">dav>"));
                webdavServer.start();
            }

            // 5) Initialize ftp server
            if (ftpServerPort > 0) {
                FtpServer ftpServer = new FtpServer("s3dav", "s3dav",
                        ftpServerPort, repository, log.getLogger(">ftp>"));
                ftpServer.start();
            }
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    private static int getPortValue(String[] args, int numParam,
            int defaultValue) {
        int result = 0;

        if (args.length >= (numParam + 1)) {
            try {
                result = Integer.parseInt(args[numParam]);
            } catch (NumberFormatException ex) {
                System.out.println("Invalid value for port number:("
                        + args[numParam] + ")");
                System.exit(1);
            }
        } else {
            result = defaultValue;
        }
        return result;
    }

    private static void createCredentials(File s3DavDir, String accessKey,
            String secretKey) {
        System.out.println("Credential creation ...");
        CredentialFactory.newCredential(s3DavDir, accessKey, secretKey);
        System.out.println("Credential created");
    }
}
